web/src/recent_senders.ts

```diff
@@ -2,60 +2,67 @@ import _ from "lodash";
 
 import {FoldDict} from "./fold_dict";
 import * as message_store from "./message_store";
+import type {Message} from "./message_store";
 import * as people from "./people";
+import type {User} from "./people";
 
 // This class is only exported for unit testing purposes.
 // If we find reuse opportunities, we should just put it into
 // its own module.
 export class IdTracker {
-    ids = new Set();
+    ids: Set<number> = new Set();
 
     // We cache the max message id to make sure that
     // typeahead code is efficient.  We don't eagerly
     // compute it, since it's plausible a spammy bot
     // could cause us to process many messages at a time
     // during fetching.
-    _cached_max_id = undefined;
+    _cached_max_id: number | undefined = undefined;
 
-    add(id) {
+    add(id: number): void {
         this.ids.add(id);
         if (this._cached_max_id !== undefined && id > this._cached_max_id) {
             this._cached_max_id = id;
         }
     }
 
-    remove(id) {
+    remove(id: number): void {
         this.ids.delete(id);
         this._cached_max_id = undefined;
     }
 
-    max_id() {
+    max_id(): number {
         if (this._cached_max_id === undefined) {
             this._cached_max_id = _.max([...this.ids]);
         }
         return this._cached_max_id || -1;
     }
 
-    empty() {
+    empty(): boolean {
         return this.ids.size === 0;
     }
 }
 
 // topic_senders[stream_id][sender_id] = IdTracker
-const stream_senders = new Map();
+const stream_senders: Map<number, Map<number, IdTracker>> = new Map();
 
 // topic_senders[stream_id][topic_id][sender_id] = IdTracker
-const topic_senders = new Map();
+const topic_senders: Map<number, FoldDict<Map<number, IdTracker>>> = new Map();
 
 // pm_senders[user_ids_string][user_id] = IdTracker
-const pm_senders = new Map();
+const pm_senders: Map<string, Map<number, IdTracker>> = new Map();
 
-export function clear_for_testing() {
+export function clear_for_testing(): void {
     stream_senders.clear();
     topic_senders.clear();
 }
 
-function max_id_for_stream_topic_sender({stream_id, topic, sender_id}) {
+function max_id_for_stream_topic_sender(opts: {
+    stream_id: number;
+    topic: string;
+    sender_id: number;
+}): number {
+    const {stream_id, topic, sender_id} = opts;
```

Comment:
You can still destructure parameters in TypeScript without extracting a separate variable:

```ts
function max_id_for_stream_topic_sender({
    stream_id,
    topic,
    sender_id,
}: {
    stream_id: number;
    topic: string;
    sender_id: number;
}): number {
```

Reply:
Yeah, I just found that more ugly and hard to read than de-structuring an `opts` variable. Happy to switch it to what you wrote if you prefer that though